Eliminate pure/elemental key words for vapor saturation related subroutines, add the vectorized versions of subroutines and change the interface for GPU computation

sjsprecious opened the pull request:
We introduce some changes to CAM codes to (1) save some computational cost and (2) make the codes portable to GPU, based on John Dennis's previous work on MG2.
- The computational cost is saved mainly due to the elimination of **pure/elemental** key words for the vapor saturation related subroutines (e.g., subroutines with name **qsat**) and using the vectorized subroutines. The major changes occur in the **wv_sat_methods.F90** and **wv_saturation.F90**, but many codes under the **physics/cam** and **chemistry** folders are also changed to accommodate the new interfaces of subroutines.
- In addition, we explicitly specify the loop index when we call the vapor saturation subroutines, which is necessary to get the code ready to be ported to GPU.
- We also make some format changes (e.g., add space) to make the coding style more consistent but this is optional.

All these changes are done based on the commit **208866b** of `cam_development` branch. We perform the following ERP test to confirm that all our changes are **BFB**.

Test suite: ERP_Ln9
Test baseline: 9-time-step CAM simulation using commit **208866b** of `cam_development` branch, making sure the tag **pumas_cam-release_v1.9** is used to check out the PUMAS Repo.
Test status: [BFB]

sjsprecious committed:
Update qsat subroutines for array input
	modified:   wv_sat_methods.F90

sjsprecious committed:
Revise the qsat related subroutines by removing the elemental and pure keywords
The major changed F90 codes locate under physics/cam, chemistry/bulk_aero, chemistry/modal_aero, chemistry/mozart, chemistry/utils
BFB change
	modified:   chemistry/bulk_aero/seasalt_model.F90
	modified:   chemistry/modal_aero/modal_aero_newnuc.F90
	modified:   chemistry/mozart/mo_gas_phase_chemdr.F90
	modified:   chemistry/utils/modal_aero_wateruptake.F90
	modified:   physics/cam/aer_rad_props.F90
	modified:   physics/cam/cam_diagnostics.F90
	modified:   physics/cam/cldfrc2m.F90
	modified:   physics/cam/cldwat.F90
	modified:   physics/cam/cldwat2m_macro.F90
	modified:   physics/cam/cloud_fraction.F90
	modified:   physics/cam/clubb_intr.F90
	modified:   physics/cam/convect_shallow.F90
	modified:   physics/cam/cospsimulator_intr.F90
	modified:   physics/cam/eddy_diff.F90
	modified:   physics/cam/hk_conv.F90
	modified:   physics/cam/macrop_driver.F90
	modified:   physics/cam/nucleate_ice_cam.F90
	modified:   physics/cam/vertical_diffusion.F90
	modified:   physics/cam/wv_sat_methods.F90
	modified:   physics/cam/wv_saturation.F90
	modified:   physics/cam/zm_conv.F90

sjsprecious committed:
Change the tag name of PUMAS to include the previous NBFB changes

	modified:   Externals_CAM.cfg

sjsprecious committed:
Change the interface for some vectorized subroutines

	modified:   micro_mg_cam.F90

sjsprecious committed:
Use vectorized qsat related subroutines
BFB changes

	modified:   ../../chemistry/bulk_aero/seasalt_model.F90
	modified:   ../../chemistry/modal_aero/modal_aero_newnuc.F90
	modified:   ../../chemistry/mozart/mo_gas_phase_chemdr.F90
	modified:   ../../chemistry/utils/modal_aero_wateruptake.F90
	modified:   aer_rad_props.F90
	modified:   cam_diagnostics.F90
	modified:   cldfrc2m.F90
	modified:   cldwat.F90
	modified:   cldwat2m_macro.F90
	modified:   cloud_fraction.F90
	modified:   convect_shallow.F90
	modified:   cospsimulator_intr.F90
	modified:   eddy_diff.F90
	modified:   hk_conv.F90
	modified:   nucleate_ice_cam.F90
	modified:   vertical_diffusion.F90
	modified:   wv_saturation.F90
	modified:   zm_conv.F90

johnmauff commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -1893,11 +1893,14 @@ real(r8) :: inv_exner_clubb(pcols,pverp)     ! Inverse exner function consistent
-        initend(:ncol,top_lev:pver))
+     do k = top_lev, pver
+        do i = 1, ncol
+           call ice_macro_tend(naai(i,k),state1%t(i,k),state1%pmid(i,k),      &
```
Do we want to create a vectorized version of ice_macro_tend?

sjsprecious replied:
Thanks John. Just push the vectorized version of ice_macro_tend.

johnmauff commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -2796,12 +2799,15 @@ real(r8) :: inv_exner_clubb(pcols,pverp)     ! Inverse exner function consistent
- 
+      do k = top_lev, pver
+         do i = 1, ncol
+            call liquid_macro_tend(npccn(i,k),state1%t(i,k), state1%pmid(i,k),    &
```
Do we want to create a vectorized version of liquid_macro_tend?

sjsprecious replied:
Thanks John. Just push the vectorized version of liquid_macro_tend.
The liquid_macro_tend and ice_macro_tend belong to CLUBB so I am not sure whether we should change them in our MG3 work.

johnmauff commented on src/physics/cam/wv_saturation.F90:
```diff
@@ -259,41 +272,41 @@ end subroutine wv_sat_final
 
 ! Compute saturation vapor pressure over ice
-elemental function svp_ice(t) result(es)
+function svp_ice(t) result(es)
```
Probably want to create a vectorized version of svp_ice

sjsprecious replied:
Thanks, it is added.

johnmauff commented on src/physics/cam/wv_saturation.F90:
```diff
@@ -259,41 +272,41 @@ end subroutine wv_sat_final
 
 ! Compute saturation vapor pressure over water
-elemental function svp_water(t) result(es)
+function svp_water(t) result(es)
```
Create a vectorized version of svp_water.

sjsprecious replied:
Thanks, it is added.

johnmauff commented on src/physics/cam/wv_saturation.F90:
```diff
@@ -259,41 +272,41 @@ end subroutine wv_sat_final
 
 ! Compute saturation vapor pressure with an ice-water transition
-elemental function svp_trans(t) result(es)
+function svp_trans(t) result(es)
```
Create a vectorized version of svp_trans

sjsprecious replied:
Thanks, it is added.

sjsprecious committed:
Update the interface of size_dist* subroutines for vectorization

	modified:   micro_mg_cam.F90

sjsprecious committed:
iAdd vectorized version of liquid_macro_tend and ice_macro_tend

	modified:   clubb_intr.F90
	modified:   macrop_driver.F90

sjsprecious committed:
Add vectorized version of svp_ice, svp_water and svp_trans

	modified:   cam_diagnostics.F90
	modified:   cldfrc2m.F90
	modified:   cloud_fraction.F90
	modified:   hetfrz_classnuc_cam.F90
	modified:   wv_sat_methods.F90
	modified:   wv_saturation.F90
	modified:   zm_microphysics.F90

sjsprecious committed:
Add vectorized version of qsat subroutines to match 2D array

	modified:   ../../chemistry/bulk_aero/seasalt_model.F90
	modified:   ../../chemistry/modal_aero/modal_aero_newnuc.F90
	modified:   ../../chemistry/mozart/mo_gas_phase_chemdr.F90
	modified:   ../../chemistry/utils/modal_aero_wateruptake.F90
	modified:   aer_rad_props.F90
	modified:   cam_diagnostics.F90
	modified:   cldwat.F90
	modified:   cldwat2m_macro.F90
	modified:   cloud_fraction.F90
	modified:   convect_shallow.F90
	modified:   cospsimulator_intr.F90
	modified:   eddy_diff.F90
	modified:   hk_conv.F90
	modified:   nucleate_ice_cam.F90
	modified:   vertical_diffusion.F90
	modified:   wv_saturation.F90
	modified:   zm_conv.F90

sjsprecious committed:
Revert a few calls to qsat* and svp* subroutins back to 1-d vectorized verision

	modified:   ../../chemistry/bulk_aero/seasalt_model.F90
	modified:   ../../chemistry/modal_aero/modal_aero_newnuc.F90
	modified:   ../../chemistry/mozart/mo_gas_phase_chemdr.F90
	modified:   ../../chemistry/utils/modal_aero_wateruptake.F90
	modified:   aer_rad_props.F90
	modified:   cam_diagnostics.F90
	modified:   cldwat.F90
	modified:   cldwat2m_macro.F90
	modified:   cloud_fraction.F90
	modified:   convect_shallow.F90
	modified:   cospsimulator_intr.F90
	modified:   eddy_diff.F90
	modified:   hetfrz_classnuc_cam.F90
	modified:   hk_conv.F90
	modified:   micro_mg_cam.F90
	modified:   nucleate_ice_cam.F90
	modified:   vertical_diffusion.F90
	modified:   zm_conv.F90
	modified:   zm_microphysics.F90

sjsprecious committed:
A minor bug fix for T dimension

	modified:   hetfrz_classnuc_cam.F90

Katetc commented:
Just a note that I changed the base for this PR from PUMAS_Development/CAM/cam_development (which we keep as a mirror to ESCOMP/CAM/cam_development) to a living branch I just made, cam_pumas_pr and will use to stage the ESCOMP/CAM/cam_development pull request. I will tag @sjsprecious in the ESCOMP/CAM PR so you can see the CAM SE review suggestions, but it is unlikely you will need to do any further work. We are also going to fold in a few other issues and code changes to this CAM PR while we have the tag, and I will need this living branch for that.
